src/Utilities/functions.jl

```diff
@@ -4,20 +4,20 @@ using Compat.Test
 """
     evalvariables(varval::Function, f::AbstractFunction)
 
-Returns the value of function `f` if each variable index `vi` is evaluate as `varval(vi)`.
+Returns the value of function `f` if each variable index `vi` is evaluated as `varval(vi)`.
 """
 function evalvariables end
 evalvariables(varval::Function, f::MOI.SingleVariable) = varval(f.variable)
 evalvariables(varval::Function, f::MOI.VectorOfVariables) = varval.(f.variables)
-evalvariables(varval::Function, f::MOI.ScalarAffineFunction) = sum(evalterm.(varval, f.terms)) + f.constant
+evalvariables(varval::Function, f::MOI.ScalarAffineFunction) = mapreduce(t->evalterm(varval, t), +, zero(f.constant), f.terms) + f.constant
```

Comment:
There are sometimes issues with `sum` not working over empty collection if I am not mistaken. Is this function tested with empty `ScalarAffineFunction` ?

Reply:
Ups, you are doing the opposite, I misread the diff, you are replacing `sum` with `mapreduce`. Is it because of the issue with empty function ?

Reply:
Yes but also `mapreduce` is faster for some reason.
```julia
julia> using BenchmarkTools

julia> a = rand(1000);

julia> f(a) = sum(i^2 for i in a)
f (generic function with 1 method)

julia> g(a) = sum(a.^2)
g (generic function with 1 method)

julia> h(a) = mapreduce((x)->(x^2), +, a)
h (generic function with 1 method)

julia> @btime h($a);
  80.205 ns (0 allocations: 0 bytes)

julia> @btime f($a);
  1.094 μs (1 allocation: 16 bytes)

julia> @btime g($a);
  691.091 ns (1 allocation: 7.94 KiB)
```

Reply:
Is it still the case in Julia v0.7 ? The allocation of `g` is also suprisingly high

Reply:
The above benchmarks are from v0.7. The allocations are not really that high at all, it is just allocating for the intermediate result of the squares. There are 1000 Float64's, each is 8 bytes, so there should be at least 7.8125 KiB of allocations. Not sure where the remaining comes from, could be some internal broadcast stuff.